process/notifier/sovereignNotifier.go

```diff
@@ -1,33 +1,205 @@
+	Marshaller          marshal.Marshalizer
+	SubscribedAddresses [][]byte
 }
 
 // NewSovereignNotifier will create a sovereign shard notifier
-func NewSovereignNotifier(config config.Config) (*sovereignNotifier, error) {
-	addresses := config.SubscribedAddresses
-	if len(addresses) == 0 {
-		return nil, errNoSubscribedAddresses
+func NewSovereignNotifier(args ArgsSovereignNotifier) (*sovereignNotifier, error) {
+	if check.IfNil(args.Marshaller) {
+		return nil, core.ErrNilMarshalizer
+	}
+	if check.IfNil(args.ShardCoordinator) {
+		return nil, errNilShardCoordinator
+	}
+
+	subscribedAddresses, err := getAddressesMap(args.SubscribedAddresses)
+	if err != nil {
+		return nil, err
 	}
 
-	log.Debug("received config", "subscribed addresses", addresses)
+	log.Debug("received config", "subscribed addresses", args.SubscribedAddresses)
 
 	return &sovereignNotifier{
-		subscribedAddresses: addresses,
+		subscribedAddresses: subscribedAddresses,
+		handlers:            make([]process.ExtendedHeaderHandler, 0),
+		mutHandler:          sync.RWMutex{},
+		headerV2Creator:     block.NewEmptyHeaderV2Creator(),
+		marshaller:          args.Marshaller,
+		shardCoordinator:    args.ShardCoordinator,
+	}, nil
+}
+
+func getAddressesMap(addresses [][]byte) (map[string]struct{}, error) {
+	numAddresses := len(addresses)
+	if numAddresses == 0 {
+		return nil, errNoSubscribedAddresses
+	}
+
+	addressesMap := make(map[string]struct{}, numAddresses)
+	for _, addr := range addresses {
+		addressesMap[string(addr)] = struct{}{}
+	}
+
+	if len(addressesMap) != numAddresses {
+		return nil, errDuplicateSubscribedAddresses
+	}
+
+	return addressesMap, nil
+}
+
+// Notify will notify the sovereign nodes about the finalized block and incoming mb txs
+// For each subscribed address, it searches if the receiver is found in transaction pool.
+// If found, IncomingMiniBlocks will contain the ordered tx hashes by execution.
+func (notifier *sovereignNotifier) Notify(outportBlock *outport.OutportBlock) error {
+	err := checkNilOutportBlockFields(outportBlock)
+	if err != nil {
+		return err
+	}
+
+	mbs, err := notifier.getAllIncomingMbs(outportBlock.TransactionPool)
+	if err != nil {
+		return err
+	}
+
+	headerV2, err := notifier.getHeaderV2(core.HeaderType(outportBlock.BlockData.HeaderType), outportBlock.BlockData.HeaderBytes)
+	if err != nil {
+		return err
+	}
+
+	extendedHeader := &block.ShardHeaderExtended{
+		Header:             headerV2,
+		IncomingMiniBlocks: mbs,
+	}
+
+	notifier.notifyHandlers(extendedHeader)
+	return nil
+}
+
+func checkNilOutportBlockFields(outportBlock *outport.OutportBlock) error {
+	if outportBlock == nil {
+		return errNilOutportBlock
+	}
+	if outportBlock.TransactionPool == nil {
+		return errNilTransactionPool
+	}
+	if outportBlock.BlockData == nil {
+		return errNilBlockData
+	}
+
+	return nil
+}
+
+func (notifier *sovereignNotifier) getAllIncomingMbs(txPool *outport.TransactionPool) ([]*block.MiniBlock, error) {
```

Comment:
I think a better name could be createAllIncomingMbs instead get...

Reply:
Refactored